src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -303,24 +307,341 @@ impl SkipWatermarkState {
     }
 }
 
+pub struct NonPkPrefixSkipWatermarkState {
+    watermarks: BTreeMap<TableId, ReadTableWatermark>,
+    remain_watermarks: VecDeque<(TableId, VirtualNode, WatermarkDirection, Datum)>,
+    compaction_catalog_agent_ref: CompactionCatalogAgentRef,
+
+    last_serde: Option<(OrderedRowSerde, OrderedRowSerde, usize)>,
+}
+
+impl NonPkPrefixSkipWatermarkState {
+    pub fn new(
+        watermarks: BTreeMap<TableId, ReadTableWatermark>,
+        compaction_catalog_agent_ref: CompactionCatalogAgentRef,
+    ) -> Self {
+        Self {
+            remain_watermarks: VecDeque::new(),
+            watermarks,
+            compaction_catalog_agent_ref,
+            last_serde: None,
+        }
+    }
+
+    pub fn from_safe_epoch_watermarks(
+        safe_epoch_watermarks: BTreeMap<u32, TableWatermarks>,
+        compaction_catalog_agent_ref: CompactionCatalogAgentRef,
+    ) -> Self {
+        let watermarks = safe_epoch_read_table_watermarks_impl(safe_epoch_watermarks);
+        Self::new(watermarks, compaction_catalog_agent_ref)
+    }
+
+    #[inline(always)]
+    pub fn has_watermark(&self) -> bool {
+        !self.remain_watermarks.is_empty()
+    }
+
+    pub fn should_delete(&mut self, key: &FullKey<&[u8]>) -> bool {
+        if let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front() {
+            let key_table_id = key.user_key.table_id;
+            let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+            match (&key_table_id, &key_vnode).cmp(&(table_id, vnode)) {
+                Ordering::Less => {
+                    return false;
+                }
+                Ordering::Equal => {
+                    let (pk_prefix_serde, watermark_col_serde, watermark_col_idx_in_pk) =
+                        match self.last_serde.as_ref() {
+                            Some(serde) => serde,
+                            None => {
+                                self.last_serde = self
+                                    .compaction_catalog_agent_ref
+                                    .watermark_serde(table_id.table_id());
+                                self.last_serde.as_ref().unwrap()
+                            }
+                        };
+                    let row = pk_prefix_serde
+                        .deserialize(inner_key)
+                        .unwrap_or_else(|_| {
+                            panic!("Failed to deserialize pk_prefix inner_key {:?} serde data_types {:?} order_types {:?}", inner_key, pk_prefix_serde.get_data_types(), pk_prefix_serde.get_order_types());
+                        });
+                    let watermark_col_in_pk = row.datum_at(*watermark_col_idx_in_pk);
+                    return direction.datum_filter_by_watermark(
+                        watermark_col_in_pk,
+                        watermark,
+                        watermark_col_serde.get_order_types()[0],
```

Comment:
Given that we use `[0]` here, does it mean that either the watermark column is always a single column or the order type for all watermark columns must be the same? Is this assumption checked and guaranteed somewhere?

Reply:
This assumes that the watermark is a single column. I've found no such guarantee on the state table side either. 

From my point of view, this should be the point that the optimizer needs to guarantee, what is the reasonable way to do it? @hzxa21 @st1page

Reply:
As said above, I think this guarantee should come from `clean_watermark_index_in_pk`. If the assumption is broken then using [0] should crash, which is fine with me.

Reply:
If there are multiple watermark column of different ordering type, I think this code won't crash but generate incorrect result.

Reply:
> the watermark column is always a single column or the order type

Yes and I think it should be assert when `watermark_col_serde` is generated?